crates/tlsn/src/prover.rs

```diff
@@ -202,95 +194,151 @@ impl Prover<state::CommitAccepted> {
-                    .into_inner();
+    /// Closes the connection from the client side.
+    pub fn client_close(&mut self) -> Result<(), ProverError> {
+        self.state
+            .tls_client
+            .client_close()
+            .map_err(ProverError::from)
+    }
 
-                // Prove tag verification of received records.
-                // The prover drops the proof output.
-                let _ = verify_tags(
-                    &mut vm,
-                    (keys.server_write_key, keys.server_write_iv),
-                    keys.server_write_mac_key,
-                    *tls_transcript.version(),
-                    tls_transcript.recv().to_vec(),
-                )
-                .map_err(ProverError::zk)?;
+    /// Closes the connection from the server side.
+    pub fn server_close(&mut self) -> Result<(), ProverError> {
+        self.state
+            .tls_client
+            .server_close()
+            .map_err(ProverError::from)
+    }
 
-                mux_fut
-                    .poll_with(vm.execute_all(&mut ctx).map_err(ProverError::zk))
-                    .await?;
+    /// Enables or disables the decryption of data from the server until the
+    /// server has closed the connection.
+    pub fn enable_decryption(&mut self, enable: bool) -> Result<(), ProverError> {
+        self.state
+            .tls_client
+            .enable_decryption(enable)
+            .map_err(ProverError::from)
+    }
 
-                let transcript = tls_transcript
-                    .to_transcript()
-                    .expect("transcript is complete");
-
-                Ok(Prover {
-                    config: self.config,
-                    span: self.span,
-                    state: state::Committed {
-                        mux_ctrl,
-                        mux_fut,
-                        ctx,
-                        vm,
-                        server_name: config.server_name().clone(),
-                        keys,
-                        tls_transcript,
-                        transcript,
-                    },
-                })
+    /// Returns if decryption of TLS traffic from the server is active.
+    pub fn is_decrypting(&self) -> Result<bool, ProverError> {
+        self.state
+            .tls_client
+            .is_decrypting()
+            .map_err(ProverError::from)
+    }
+
+    /// Polls the prover to make progress.
+    pub fn poll(&mut self, cx: &mut Context) -> Poll<Result<(), ProverError>> {
+        let _ = self.state.mux_fut.poll_unpin(cx)?;
+
+        match self.state.tls_client.poll(cx) {
+            Poll::Ready(output) => {
+                self.state.output = Some(output?);
+                Poll::Ready(Ok(()))
             }
-            .instrument(span)
-        });
+            Poll::Pending => Poll::Pending,
+        }
+    }
+
+    /// Returns a committed prover after the TLS session has completed.
+    pub fn finish(self) -> Option<Prover<state::Committed>> {
```

Comment:
```suggestion
    pub fn finish(self) -> Result<Prover<state::Committed>, ProverError> {
```

so we can return the reason why the state transition failed